tatanka/db/orderbook.go

```diff
@@ -0,0 +1,216 @@
+// This code is available on the terms of the project LICENSE.md file,
+// also available online at https://blueoakcouncil.org/license/1.0.0.
+
+package db
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	"decred.org/dcrdex/dex/encode"
+	"decred.org/dcrdex/dex/lexi"
+	"decred.org/dcrdex/tatanka/tanka"
+)
+
+type OrderUpdate struct {
+	Sig []byte `json:"sig"`
+	*tanka.Order
+}
+
+func (o *OrderUpdate) MarshalBinary() ([]byte, error) {
+	const orderVer = 0
+	var b encode.BuildyBytes = make([]byte, 1, 32+4+4+1+8+8+8+8+8+8+4+8+len(o.Sig))
+	b[0] = orderVer
+	sell := encode.ByteFalse
+	if o.Sell {
+		sell = encode.ByteTrue
+	}
+	b = b.AddData(o.From[:]).
+		AddData(encode.Uint32Bytes(o.BaseID)).
+		AddData(encode.Uint32Bytes(o.QuoteID)).
+		AddData(sell).
+		AddData(encode.Uint64Bytes(o.Qty)).
+		AddData(encode.Uint64Bytes(o.Rate)).
+		AddData(encode.Uint64Bytes(o.LotSize)).
+		AddData(encode.Uint64Bytes(o.MinFeeRate)).
+		AddData(encode.Uint64Bytes(uint64(o.Stamp.UnixMilli()))).
+		AddData(encode.Uint32Bytes(o.Nonce)).
+		AddData(encode.Uint64Bytes(uint64(o.Expiration.UnixMilli()))).
+		AddData(encode.Uint64Bytes(o.Settled)).
+		AddData(o.Sig)
+
+	return b, nil
+}
+
+func (o *OrderUpdate) UnmarshalBinary(b []byte) error {
+	if o == nil {
+		return errors.New("nil order update")
+	}
+	const orderVer = 0
+	o.Order = new(tanka.Order)
+	ver, pushes, err := encode.DecodeBlob(b, 13)
+	if err != nil {
+		return fmt.Errorf("error decoding order update blob: %w", err)
+	}
+	if ver != orderVer {
+		return fmt.Errorf("unknown order update version %d", ver)
+	}
+	if len(pushes) != 13 {
+		return fmt.Errorf("unknown number of order update blob pushes %d", len(pushes))
+	}
+	copy(o.From[:], pushes[0])
+	o.BaseID = encode.BytesToUint32(pushes[1])
+	o.QuoteID = encode.BytesToUint32(pushes[2])
+	o.Sell = pushes[3][0] == encode.ByteTrue[0]
+	o.Qty = encode.BytesToUint64(pushes[4])
+	o.Rate = encode.BytesToUint64(pushes[5])
+	o.LotSize = encode.BytesToUint64(pushes[6])
+	o.MinFeeRate = encode.BytesToUint64(pushes[7])
+	o.Stamp = time.UnixMilli(int64(encode.BytesToUint64(pushes[8])))
+	o.Nonce = encode.BytesToUint32(pushes[9])
+	o.Expiration = time.UnixMilli(int64(encode.BytesToUint64(pushes[10])))
+	o.Settled = encode.BytesToUint64(pushes[11])
+	o.Sig = pushes[12]
+
+	return nil
+}
+
+type OrderBooker interface {
+	OrderIDs(fromIdx, nOrders uint64) ([]tanka.ID32, bool, error)
+	Order(id tanka.ID32) (*OrderUpdate, error)
+	Orders(id []tanka.ID32) ([]*OrderUpdate, error)
+	FindOrders(filter *OrderFilter) ([]*OrderUpdate, error)
+	Add(*OrderUpdate) error
+	Update(*OrderUpdate) error
+	Delete(id tanka.ID32) error
+}
+
+type OrderFilter struct {
+	IsSell *bool
+	Check  func(*OrderUpdate) (ok, done bool)
+}
+
+var _ OrderBooker = (*OrderBook)(nil)
+
+type OrderBook struct {
+	orderBook            *lexi.Table
+	orderBookOrderIDIdx  *lexi.Index
+	orderBookStampIdx    *lexi.Index
+	orderBookSellRateIdx *lexi.Index
+}
```

Comment:
Do orders need to be stored to disk?

Reply:
If there were a lot of orders, would you store them all in memory?

Reply:
An order is roughly 100 bytes
```Go
type Order struct {
	From    PeerID // 32 bytes
	BaseID  uint32 // 4
	QuoteID uint32 // 4
	Sell    bool   // 1
	Qty     uint64 // 8
	Rate    uint64 // 8
	LotSize    uint64    // 8
	Stamp      time.Time // 24, I think
	Expiration time.Time // 24
}
```
Let's say each market has 1000 orders per side. Then a market's footprint in memory will be `100 * 1000 * 2 bytes = 200 kB`.

Reply:
Ok, no db just a map then. I'll rework this.